### PR TITLE
Obtain SoC part number from the device tree and not Kconfig

### PR DESCRIPTION
**Pull request by mnkp**
Virtually all of the SoCs require part number information to compile the code. Originally this information was provided by a Kconfig option. After introduction of the DT some part number dependent configuration settings such as RAM and flash size were moved to DT. Consequently we need to maintain and set part number information twice, once in Kconfig and once in DT.

This PR adds DT bindings to allow to obtain SoC part number from the device tree. The corresponding Kconfig options can be removed.

Zephyr DT is supposed to be compatible with Linux DT, however Linux DT does not seem to contain part number information. Following proposal is Zephyr specific.

Since the DT root node is supposed to represent the board I propose to attach the part-number property to the soc node. The proposal needs to be carefully reviewed.

### CODE REVIEW

**Comment by galak** [comment]
Is it possible to do:


```
soc {
	compatible = "silabs,EFR32FG1P133F256GM48";
...
```

**Comment by mnkp** [comment]
> Is it possible to do:
> 
> ```
> soc {
> 	compatible = "silabs,EFR32FG1P133F256GM48";
> ...
> ```

We could do 
```
	compatible = "silabs,EFR32FG1P133F256GM48", "soc,silabs";
```
maybe also
```
	compatible = "silabs,EFR32FG1P133F256GM48", "soc,silabs", "soc";
```
I'm not really sure that we need soc,silabs.yaml binding (in case of SiLabs chips) since we do not differentiate it with the general soc and most likely we will not in the future. It's also unlikely that other SoC families will need to use vendor specific properties. Nevertheless, I kept the design consistent with what we do for other DT nodes.

**Comment by erwango** [comment]
@mnkp,

> however Linux DT does not seem to contain part number information.

This exists on some parts:
https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/stm32f429-disco.dts#L55

I definitely agree this is where we need to go (and potentially get rid of most `soc/` code).
Having the info inside SoC dtsi files also makes sense. Let's see how we can make best use of it.

**Comment by mnkp** [comment]
> Is it possible to do:
> 
> ```
> soc {
> 	compatible = "silabs,EFR32FG1P133F256GM48";
> ...
> ```

Looking at some old discussions on Slack I think I may have misunderstood @galak's proposal. And @erwango may have had the same in mind.

The alternative approach would be not to add 'part-number' property but rather generate a define for 'compatible' property. That has some advantages. An obvious one is that 'compatible' is a standard property. 

In the nearby future we will want to get rid of CONFIG_SOC_SERIES and CONFIG_SOC_FAMILY Kconfig options. The 'compatible' property approach could be expanded to support it. We would need to enforce the second and third string of the property to be the name of the series and the name of the family. 
```
	compatible = "silabs,EFR32FG1P133F256GM48", "silabs,efr32fg1p", "silabs,exx32";
```
There is one disadvantage. The 'compatible' property is build around notion of selecting the best matching model from the list provided. In our case we would overload that meaning. The list wouldn't be optional but strictly predetermined. Skipping one of the elements would break compilation. That's not how compatible is supposed to work.

One other minor issue to keep in mind, the part number is mostly required to select a vendor provided include header, directory or pass as a preprocessor define. It _must_ have a form of a string containing the part name with no decorations. e.g. "EFR32FG1P133F256GM48" and not "silabs,EFR32FG1P133F256GM48". Since cmake supports string processing operations the part number could be extracted from the property. Without cmake however that property wouldn't be of much value.

So, is using a 'compatible' property a better way to go?

**Comment by pabigot** [comment]
@mnkp can you link to the Slack discussions?

Absent seeing that, I like putting it in `compatible` rather than creating a new property name.

I don't see an obvious motivation for compatible with `"soc,silabs"` or `"soc"`.  Something like this *might* make sense: 
```
soc {
	compatible = "silabs,EFR32FG1P133F256GM48", "silabs,EFR32FG", "silabs,EFR32";
}
```

If the tooling generated this as:
```
#define DT_SOC_COMPATIBLE_SILABS_EFR32FG1P133F256GM48 1
#define DT_SOC_COMPATIBLE_SILABS_EFR32FG 1
#define DT_SOC_COMPATIBLE_SILABS_EFR32 1
```
would that handle things?  That'd probably be enough for a generic header to produce a define with the correct string (selecting from a set of known values using `#ifdef`).

Given the (proposed) need for multiple strings in the `compatible` value I'm not clear how to automatically generate a define with a string value.  Maybe:
```
compatible = "soc,EFR32FG1P133F256GM48"
```
and special-case the processing of any `compatible` value starting with `"soc,"` to generate the remainder:
```
#define DT_SOC_COMPATIBLE "EFR32FG1P133F256GM48"
```
though that's a little ugly.

**Comment by pabigot** [comment]
tl;dr: There's a lot of variation in how vendors express the [hardware configuration hierarchy](https://docs.zephyrproject.org/latest/porting/board_porting.html#hardware-configuration-hierarchy). I think the overall goal of this PR is important, but that a successful solution requires a thorough review of existing practice and may require some vendors to change the way they do things today.

Previous slack discussion involving @galak and @mbolivar is [here](https://zephyrproject.slack.com/archives/C18PLHN8H/p1544809052942100).

Stepping back, my understanding is that before device-tree Kconfig would generate defines like:
```
CONFIG_SOC_FAMILY_EXX32=y
CONFIG_SOC_SERIES_EFR32MG12P=y
CONFIG_SOC_PART_NUMBER_EFR32MG12P332F1024GL125=y
```
where the first two come from `board_defconfig` and the last from a `select` clause in `Kconfig.board`.  Silicon Labs and NXP follow this model; Nordic and Intel follow a model that produces:
```
CONFIG_SOC_SERIES_QUARK_SE=y
CONFIG_SOC_FAMILY_QUARK=y
CONFIG_SOC_QUARK_SE_C1000=y
```
where all three come from `board_defconfig` and there is no `_PART_NUMBER` component in the identifier.

With device tree you want a way to instead generate:
```
DT_SOC_FAMILY_EXX32=y
DT_SOC_SERIES_EFR32MG12P=y
DT_SOC_PART_NUMBER_EFR32MG12P332F1024GL125=y
```
from properties located in a device tree node, probably at `/soc`.

Is this an accurate summary of the problem?

You (@mnkp) also said:

> The name of the SoC series and SoC family has to correspond strictly to the names of folders we use to keep relevant SoC files.

This doesn't seem quite correct as the path for above is `soc/arm/silabs_exx32/efr32mg12p` which requires a prefix `silabs_` that isn't part of any existing symbol.  It's not obvious where that's getting injected.

You've also said this could be zephyr-specific, but @erwango pointed out existing practice in Linux, specifically for the `compatible` property.

So assuming that's all accurate: I don't like either `soc,silabs` or `silabs,soc` as `compatible` values: they both go against the standard of `manufacturer,model` for that property (the Linux example for `st,stm32f429` conforms).

Looking at the current contents of the `soc` directory I would propose something like the following:
```
soc {
    compatible = "silabs,exx32";
    series = "efr32mg12p";
    part-number = "efr32mg12p332f1024gl125";
};

soc {
    compatible = "nordic,nrf"; // currently nordic,nrf52840{,-qiaa}
    series = "nrf52";           // currently nrf52x
    part-number = "nrf52840-qiaa";
};

soc {
    compatible = "intel,quark";  // currently intel,quark_se_c1000; atom and other families under x86 are anomalous
    series = "quark_se";
    part-number = "quark_se_c1000";
};
```

The `compatible` value should identify the `soc/cpu/manufacturer_family` file path root (some like `atom` would have to be renamed).  The last component of `compatible` could be used by the extractdts infrastructure to synthesize `DT_SOC_FAMILY_model`.  Alternatively a `family` property could be introduced and `compatible` left alone.  The `series` and `part-number` properties would translate directly with a synthesized prefix of `DT_SOC_SERIES_` and `DT_SOC_PART_NUMBER_` respectively.

Unifying the way that family, series, and part are expressed across vendors may require changes for some vendors in preprocessor identifier, file system hierarchy, or both.  Also the existing `compatible` properties at the `soc` node have wide variation including values like `simple-bus` which would have to be reviewed.

**Comment by galak** [comment]
So I think compatiable at the soc node makes sense, we can use the work I've got going PR #12108

**Comment by pabigot** [comment]
@mnkp Is this PR still relevant?

**Comment by mnkp** [comment]
> Is this PR still relevant?

It is unlikely that the solution proposed in this PR will be adopted as the one that got more traction was the proposal from @galak 

> So I think compatiable at the soc node makes sense, we can use the work I've got going PR #12108

There was some initial work going on in #12428 but the PR got closed. Using compatible seems a more proper approach, inline with the idea of DTS, but it's not quite as easy as having a dedicated part-number property.

@galak could you quickly revisit this PR and close it if it's no more relevant. The discussion around generating part number from DTS was never finalized.

**Comment by nashif** [comment]
@galak  ping

**Comment by dleach02** [comment]
@galak, @mnkp, Can we close this PR?

**Comment by nashif** [comment]
stale for a long time already..
